ts/user-picker/user-picker.controller.ts

```diff
@@ -0,0 +1,342 @@
+							this.$scope.users.splice(myIndex, 1);
+							this.$scope.users.unshift(me);
+						} else {
+							this.userPickerService.getMe().then((me: IUserLookup) => {
+								this.tidyUp([me]).then((meComplete: IUserLookup[]) => {
+									this.$scope.users.unshift(meComplete[0]);
+								});
+							});
+						}
+					}
+				}
+			});
+
+			this.$scope.$watch("showFormerEmployees", (newValue: boolean, oldValue: boolean) => {
+				if (!!this.$scope.showFormerEmployees && newValue !== oldValue) {
+					this.resetUsers();
+					this.refresh();
+				}
+			});
+
+			this.$scope.$watchCollection("bypassOperationsFor", (newValue: number[], oldValue: number[]) => {
+				if (newValue !== undefined) {
+					this.userPickerService.getUsersByIds(newValue).then((bypassedUsers: IUserLookup[]) => {
+						this.tidyUp(bypassedUsers).then((completedByPassedUsers: IUserLookup[]) => {
+							_.each(completedByPassedUsers, (byPassedUser: IUserLookup) => {
+								if (_.find(this.$scope.users, (user: IUserLookup) => { return user.id === byPassedUser.id; }) === undefined) {
+									this.$scope.users.push(byPassedUser);
+								}
+							});
+						});
+					});
+				}
+			});
+
+			this.$scope.$watchGroup(["appId", "operations"], (newValue: [number, string[]], oldValue: [number, string[]]) => {
+				if (angular.isDefined(newValue) && angular.isDefined(newValue[0]) &&
+					angular.isDefined(newValue[1]) && newValue[1].length > 0 &&
+					newValue[0] != oldValue[0] && !_.isEqual(newValue[1], oldValue[1])) {
+					this.resetUsers();
+					this.refresh();
+				}
+			});
+
+			this.$scope.find = (search: string): void => {
+				this.resetUsers();
+				this.refresh(search);
+			};
+
+			this.$scope.loadMore = (): void => {
+				if (this.$scope.displayAllUsers === undefined || this.$scope.displayAllUsers === false) {
+					this.$scope.lastPagingOffset += MAGIC_PAGING;
+					this.$scope.loadingMore = true;
+					this.refresh().then(() => { this.$scope.loadingMore = false; });
+				}
+			};
+
+			this.$scope.onSelectedUserChanged = (user: IUserLookup): void => {
+				this.setViewValue(user);
+				if (!!this.$scope.onSelect()) {
+					this.$scope.onSelect();
+				}
+			};
+
+			this.$scope.onSelectedUsersChanged = (): void => {
+				this.setViewValue(this.$scope.selectedUsers);
+				if (!!this.$scope.onSelect()) {
+					this.$scope.onSelect();
+				}
+			};
+
+			this.$scope.onSelectedUserRemoved = (): void => {
+				this.setViewValue(this.$scope.selectedUsers);
+				if (!!this.$scope.onRemove()) {
+					this.$scope.onRemove();
+				}
+			};
+		}
+
+		/**
+		 * Fetches most of the additional properties given in the attributes of the directive (custom-info, customInfoAsync, etc).
+		 * Also handles homonyms.
+		 */
+		private tidyUp(users: IUserLookup[]): ng.IPromise<IUserLookup[]> {
+			let promises = new Array<ng.IPromise<any>>();
+			let customInfoDico: { [userId: number]: ng.IPromise<string> } = {};
+			let homonymsDico = {};
```

Comment:
It's actually a Dictionary<int, int>, where the key represents the user's id and the value the index of the promise in the 'promises' array, but okay, I'll add a type def